// @meteorwallet/sdk/dist/meteor-common/src/types/NearCommonTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ENearAccessKeyPermission = exports.ENearTokenWrapId = exports.ENearNetwork = void 0;
var ENearNetwork;
(function (ENearNetwork) {
    ENearNetwork["localnet"] = "localnet";
    ENearNetwork["testnet"] = "testnet";
    ENearNetwork["betanet"] = "betanet";
    ENearNetwork["mainnet"] = "mainnet";
})(ENearNetwork = exports.ENearNetwork || (exports.ENearNetwork = {}));
var ENearTokenWrapId;
(function (ENearTokenWrapId) {
    ENearTokenWrapId["WRAP_TESTNET"] = "wrap.testnet";
    ENearTokenWrapId["WRAP_MAINNET"] = "wrap.mainnet";
})(ENearTokenWrapId = exports.ENearTokenWrapId || (exports.ENearTokenWrapId = {}));
var ENearAccessKeyPermission;
(function (ENearAccessKeyPermission) {
    ENearAccessKeyPermission["FULL_ACCESS"] = "FULL_ACCESS";
    ENearAccessKeyPermission["FUNCTION_CALL"] = "FUNCTION_CALL";
})(ENearAccessKeyPermission = exports.ENearAccessKeyPermission || (exports.ENearAccessKeyPermission = {}));
